lib/kauth/login.go

```diff
@@ -0,0 +1,121 @@
+package kauth
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/astore/common"
+	"github.com/enfabrica/enkit/astore/rpc/auth"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"github.com/enfabrica/enkit/lib/logger"
+	"github.com/enfabrica/enkit/lib/retry"
+	"github.com/enfabrica/enkit/lib/srand"
+	"github.com/pkg/browser"
+	"golang.org/x/crypto/nacl/box"
+	"golang.org/x/crypto/ssh"
+	"math/rand"
+)
+
+func init() {
+	browser.Stdout = nil
+	browser.Stderr = nil
+}
+
+type EnkitCredentials struct {
+	Token string
+	// The below fields can be possibly empty if the auth server does not support CA certificates.
+	CaHosts        []string
+	CAPublicKey    string
+	PrivateKey     ed25519.PrivateKey
+	SSHCertificate *ssh.Certificate
+}
+
+// PerformLogin will login with the provider auth client, retry and logger. It does not care about the cache.
+// If you wish to save the result, please call SaveCredentials
+func PerformLogin(authClient auth.AuthClient, l logger.Logger, repeater *retry.Options, username, domain string) (*EnkitCredentials, error) {
+	rng := rand.New(srand.Source)
+	pubBox, privBox, err := box.GenerateKey(rng)
+	if err != nil {
+		return nil, err
+	}
+	sshPub, sshPriv, err := kcerts.MakeKeys(kcerts.GenerateED25519)
```

Comment:
this is not needed until much later, move it down closer to where it is used?
makes the code harder to follow.